machineconfiguration/v1alpha1/types_osimagestream.go

```diff
@@ -0,0 +1,124 @@
+// for the MachineConfigPools to be used as base OS images.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=osimagestreams,scope=Cluster
+// +kubebuilder:subresource:status
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/2555
+// +openshift:file-pattern=cvoRunLevel=0000_80,operatorName=machine-config,operatorOrdering=01
+// +openshift:enable:FeatureGate=OSStreams
+// +kubebuilder:metadata:labels=openshift.io/operator-managed=
+// +kubebuilder:validation:XValidation:rule="self.metadata.name == 'cluster'",message="osimagestream is a singleton, .metadata.name must be 'cluster'"
+type OSImageStream struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec contains the desired OSImageStream config configuration.
+	// +required
+	Spec *OSImageStreamSpec `json:"spec,omitempty"`
+
+	// status describes the last observed state of this OSImageStream.
+	// Populated by the MachineConfigOperator after reading release metadata.
+	// When not present, the controller has not yet reconciled this resource.
+	// +optional
+	Status *OSImageStreamStatus `json:"status,omitempty"`
+}
+
+// OSImageStreamStatus describes the current state of a OSImageStream.
+// +kubebuilder:validation:XValidation:rule="!has(self.availableStreams) || size(self.availableStreams) == 0 || (has(self.defaultStream) && size(self.defaultStream) != 0)",message="defaultStream must be set when availableStreams is not empty"
+// +kubebuilder:validation:XValidation:rule="!has(self.defaultStream) || self.defaultStream in self.availableStreams.map(s, s.name)",message="defaultStream must reference a stream name from availableStreams"
+type OSImageStreamStatus struct {
+	// availableStreams is a list of the available OS Image Streams
+	// available and their associated URLs for both OS and Extensions
+	// images.
+	//
+	// A maximum of 100 streams may be specified.
+	// +optional
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=100
+	// +listType=map
+	// +listMapKey=name
+	AvailableStreams []OSImageStreamURLSet `json:"availableStreams,omitempty"`
+
+	// defaultStream is the name of the stream that should be used as the default
+	// when no specific stream is requested by a MachineConfigPool.
+	// Must reference the name of one of the streams in availableStreams.
+	// Must be set when availableStreams is not empty.
+	// When not set and availableStreams is empty, controllers should use the default one stated in the release image.
```

Comment:
```suggestion
	// When not set and availableStreams is empty, controllers should use the default one stated in the release image.
	// Must not exceed 70 characters in length.
	// Must only contain alphanumeric characters, hyphens ('-'), or dots ('.').
```

We may also want to clarify on the "why's" of those constraints.

Reply:
Done. I've not removed the original `Must only contain alphanumeric characters, hyphens ('-'), or dots ('.').` for consistency reasons, despite I have introduced a sentence with the explanation that duplicates part of the message.

Reply:
It does not mention that is needs to be between 1 and 70 characters from what I can see.